### DISCUSSION TITLE
On-demand getter pattern best practice

### DISCUSSION THREAD

**Discussion by scpedicini** [pmndrs/zustand#1166] [Q&A]
My store is going to have a number of fields and corresponding populateField methods. They may be called numerous times, across numerous components. Normally when I build things like this, I prefer to have a public on-demand getter which is backed by a hidden field along the lines of:

```ts
private string _userName;
public async string UserName 
{
    get {
        if (!_userName) {
            _userName = await fetch('/username');
        }
        return _userName;
    }
}
```

I'm not sure how to duplicate something like this in Zustand. I thought about having a special Symbol for a state of FETCHING that any zustand field could be set to while it is in a fetch mode, but than my typescript definitions get unwieldy since I would have something like:

```ts
interface MyStore {
    userName: string | null | Symbol;
}
```

Additionally, any component using userName would have to account for this "fetching state" as well, instead of just preferring to look at "null" or string.

Than I settled on a dictionary `dataState` that would tracking the "fetch" state for any given zustand field:
```ts
enum FetchState {
    Uninitialized = 0,
    Fetching = 1,
    Complete = 2,
    Error = 3
}

interface GlobalStore {
    userName: string | null;
    populateUserName: () => void;

    // string is the name of the backing field
    dataState: Record<string, FetchState>
}

const useStore = create<GlobalStore>()((set, get) => ({
    dataState: {
        'userName': FetchState.Uninitialized
    },    
    userName: null,

    // this should be run by any component that needs to know the userName
    populateUserName: async () => {
        const _dataState = get().dataState;

        if (_dataState['userName'] === FetchState.Uninitialized) {
            _dataState['userName'] = FetchState.Fetching; 
            
            // Anyway to avoid stuffing ENTIRE dictionary back in, and just updating a single key-value pair?
            set(({ dataState: _dataState })); // possible atomicity issues? 

            const response = await fetch('http://localhost:20005/user');

            if (response.ok) {
                const serverResponse = await response.json();
                const _userName = serverResponse.data.name;
                _dataState['userName'] = FetchState.Complete;
                set(({
                    userName: _userName,
                    dataState: _dataState
                }));
            } else {
                _dataState['userName'] = FetchState.Error;
                // do I need to use the get() again to get the "most current" version of the dataState
                // or can I continue to use _dataState throughout this function as I am doing?
                set(({ dataState: _dataState }));
            }
        }
    }
}));
```

Every component that relied on a field (in this case, `userName`), would just need to make sure that they preemptively called `populateUserName` at the beginning of their component.

```tsx
function Dashboard() {
    const userName = useStore((state) => state.userName);
    const populateUserName = useStore((state) => state.populateUserName);

    populateUserName(); // should this be inside a useEffect() or is this okay?

    return (
        <h1>Welcome {userName}</h1> 
    )
}
```

Is there a better way to achieve something like this? It is kind of verbose especially since I'll need to do this at a per-field level (profilePreferences, populateProfilePreferences, etc).

**Comment by dai-shi** [comment]
If I understand the requirement correctly, true on-demand getter is not possible with zustand, or maybe even with react hooks.

I think the simple fetching pattern is something like this:
```js
const useFetchStore = create((set, get) => ({
  state: 'uninitialized',
  data: null,
  startFetch: async () => {
    if (get().state !== 'uninitialized') {
      return
    }
    set({ state: 'fetching' })
    try {
      const data = await fetch(...).then((res) => res.json())
      set({ state: 'done', data })
    } catch (e) {
      set({ state: 'error' })
    }
  },
}))
```

**Reply by scpedicini** [reply]
That seems pretty similar to what I wrote, the only difference is I'm trying to account for multiple fields any of which can be separately fetched on demand. that's why I thought using a dictionary that could track the fetch state of any given field by name within the store would be the best approach?

**Reply by dai-shi** [reply]
You can use a dictionary object, but if it's in a store, you want to avoid mutating it. https://github.com/pmndrs/zustand/blob/main/docs/immutable-state-and-merging.md

**Reply by scpedicini** [reply]
> https://github.com/pmndrs/zustand/blob/main/docs/immutable-state-and-merging.md

@dai-shi I think that's what I'm doing, right? According to what I've read in the docs, using the `set` command is already shorthand for an immutable pattern.

In the `populateUserName` command, I am using `get()` to pull the dictionary `dataState` down though. I then make modifications to the dictionary, and then stuff it back in using:
```ts
set({ dataState: _dataState })
```

Am I doing this wrong?

**Reply by dai-shi** [reply]
```js
_dataState['userName'] = FetchState.Fetching;
```
That ☝️  is mutating the object.
Immutable updating will be something like this 👇 :
```js
set({ dataState: { ..._dataState, userName:  FetchState.Fetching } })
```

**Reply by scpedicini** [reply]
@dai-shi thanks, I've grabbed the `immer` package, and changed the code to the following. Does this follow the immutable pattern?

```ts
populateUserName: async () => {
    if (get().dataState['userName'] === FetchStates.Uninitialized) {

        set(produce(state => {
            state.dataState['userName'] = FetchStates.Fetching;
        }));

        const response = await fetch('http://localhost:20005/user');

        let success = false;

        if (response.ok) {
            const serverResponse = await response.json();
            const _userName = serverResponse?.data?.name;
            if (_userName) {
                set(produce(state => {
                    state.userName = _userName;
                    state.dataState['userName'] = FetchStates.Complete;
                }));
                success = true;
                console.log('populateUserName: success');
            }
        }

        if (!success) {
            set(produce(state => {
                state.dataState['userName'] = FetchStates.Error;
            }));
        }
    }
}
```

I'm still modifying it partially, but this is through the recommended `produce` pattern for reducing nested structures.

**Reply by dai-shi** [reply]
Yeah, immer is a solution too.
